web/treadmill/treadmill.py

```diff
@@ -0,0 +1,172 @@
+                'Public',
+                'Random',
+                'Public Identity',
+                'Random Identity',
+            )[advertisement.address.address_type]
+            self.rssi = advertisement.rssi
+            self.data = advertisement.data.to_string('\n')
+
+    def __init__(self, hci_source, hci_sink):
+        super().__init__()
+        random_address = Address.generate_static_address()
+        self.device = Device.with_hci('Bumbleton', random_address, hci_source, hci_sink)
+        self.scan_entries = {}
+        self.listeners = {}
+        self.peer = None
+        self.device.on('advertisement', self.on_advertisement)
+        self.device.on('connection', self.on_connection)
+
+    async def start(self):
+        print('### Starting Scanner')
+        self.scan_entries = {}
+        self.emit_scanning_update()
+        await self.device.power_on()
+        await self.device.start_scanning()
+        print('### Scanner started')
+
+    async def stop(self):
+        # TODO: replace this once a proper reset is implemented in the lib.
+        await self.device.host.send_command(HCI_Reset_Command())
+        await self.device.power_off()
+        print('### Scanner stopped')
+
+    def emit_scanning_update(self):
+        if listener := self.listeners.get('scanning_updates'):
+            listener(list(self.scan_entries.values()))
+
+    def emit_hr_update(self, value: int, time: str):
+        if listener := self.listeners.get('hr_updates'):
+            listener({"value": value, "time": time})
+
+    def emit_connection_updates(self, connection: Connection):
+        if listener := self.listeners.get('connection_updates'):
+            listener(
+                {
+                    'handle': connection.handle,
+                    'role': connection.role_name,
+                    'self_address': str(connection.self_address),
+                    'peer_address': str(connection.peer_address),
+                    'is_encrypted': "Yes" if connection.is_encrypted else "No",
+                }
+            )
+
+    def emit_on_security_request(self):
+        if listener := self.listeners.get('on_security_request'):
+            listener()
+
+    def on(self, event_name, listener):
+        self.listeners[event_name] = listener
+
+    def on_advertisement(self, advertisement: Advertisement):
+        uuids = advertisement.data.get(
+            AdvertisingData.COMPLETE_LIST_OF_16_BIT_SERVICE_CLASS_UUIDS,
+        )
+        if not uuids:
+            return
+        if not GATT_HEART_RATE_SERVICE in uuids:
+            return
+
+        self.scan_entries[advertisement.address] = self.ScanEntry(advertisement)
+        self.emit_scanning_update()
+
+    @AsyncRunner.run_in_task()
+    async def on_connection(self, connection: Connection):
+        self.emit_connection_updates(connection)
+        connection.listener = self
+        connection.on('security_request', self.on_security_request)
```

Comment:
If you want to prompt the user to accept/reject pairing, you should probably use the `PairingDelegate` mechanism rather than intercepting security requests. See `apps/pair.py` for an example (see how the `--prompt` option is handled, it does exactly that: ask the user to accept pairing or not).